language/diem-vm/src/diem_transaction_executor.rs

```diff
@@ -404,22 +405,22 @@ impl DiemVM {
 
     fn read_writeset(
         &self,
-        remote_cache: &StateViewCache<'_>,
+        state_view: &dyn StateView,
```

Comment:
`&impl` instead of `&dyn`?

Reply:
This is not possible unfortunatelly. The reason is that the api between `diem-vm` and `executor` takes a `dyn StateView` as input, thus the actual type of this StateView is not inferrable.

Reply:
Since `dyn Trait` implements `Trait`, an argument of type `dyn Trait` is compatible with a parameter of type `impl Trait`. (`fn foo(val: impl Trait)` is just a sugar for `fn foo<T: Trait>(val: T)`). Using dyn here is not only overly restrictive but also comes with a performance hit.

I would go as far as arguing that `dyn` types should be avoided unless there are specific needs for dynamic dispatch, 
and that most `dyn Trait`s in our repo don't satisfy this condition and should therefore be converted to `impl Trait`s instead. 

Going one step further, I actually suspect that most (if not all) valid usages of `dyn` can eventually be reduced to the need for some form of heterogeneous collections. Here are a few examples:
- `Vec<dyn Trait>`
- Passing a Rust closure as a callback to a C function
  - One type being able to represent different types of closures
- Implementing Haskell-ish lazy evaluation
  - Storing different types of closures in your lazily evaluated AST